src/i2c_bit_bang.cpp

```diff
@@ -0,0 +1,255 @@
+}
+
+void i2c_bit_bang::send_stop_condition()
+{
+  m_sda->level(false);
+
+  m_scl->level(true);
+  high_speed_delay(m_scl_high_ticks);
+  m_sda->level(true);
+  high_speed_delay(m_scl_high_ticks);
+}
+
+void i2c_bit_bang::write_address(hal::byte p_address,
+                                 function_ref<hal::timeout_function> p_timeout)
+{
+  // write the address
+  auto acknowledged = write_byte(p_address, p_timeout);
+
+  if (!acknowledged) {
+    hal::safe_throw(hal::no_such_device((p_address >> 1), this));
+  }
+}
+
+void i2c_bit_bang::write(std::span<const hal::byte> p_data_out,
+                         function_ref<hal::timeout_function> p_timeout)
+{
+  bool acknowledged;
+  for (const hal::byte& data : p_data_out) {
+
+    acknowledged = write_byte(data, p_timeout);
+
+    if (!acknowledged) {
+      hal::safe_throw(hal::io_error(this));
+    }
+  }
+}
+
+bool i2c_bit_bang::write_byte(hal::byte p_byte_to_write,
+                              function_ref<hal::timeout_function> p_timeout)
+{
+  using namespace std::chrono_literals;
+  constexpr auto byte_length = 8;
+  constexpr hal::byte bit_select = 0b0000'0001;
+  hal::byte bit_to_write = 0;
+  uint32_t i;
+  for (i = 0; i < byte_length; i++) {
+
+    bit_to_write = static_cast<hal::byte>(
+      (p_byte_to_write >> (byte_length - (i + 1))) & bit_select);
+
+    write_bit(bit_to_write, p_timeout);
+  }
+
+  // look for the ack
+  auto ack_bit = read_bit();
+  // if ack bit is 0, then it was acknowledged (true)
+  return (ack_bit == 0);
+}
+/*
+   for writing a bit you want to make set the data line first, then toggle the
+   level of the clock then check if the level was indeed toggled or if the
+   peripheral is stretching the clock. After this is done, you are able to set
+   the clock back low.
+*/
+void i2c_bit_bang::write_bit(hal::byte p_bit_to_write,
+                             function_ref<hal::timeout_function> p_timeout)
+{
+  m_sda->level(static_cast<bool>(p_bit_to_write));
+  m_scl->level(true);
+  high_speed_delay(m_scl_high_ticks);
+
+  // if scl is still low after we set it high, then the peripheral is clock
+  // stretching
+  while (m_scl->level() == 0) {
+    p_timeout();
+  }
+
+  m_scl->level(false);
+  high_speed_delay(m_scl_low_ticks);
+}
+
+void i2c_bit_bang::read(std::span<hal::byte> p_data_in,
+                        function_ref<hal::timeout_function> p_timeout)
+{
+  uint32_t size_of_span = p_data_in.size(), i = 0;
+  for (hal::byte& data : p_data_in) {
+    data = read_byte();
+    i++;
+
+    if (i < size_of_span) {
+      // if the iterator isn't done, then we ack whatever data we read
+      write_bit(0, p_timeout);
+
+    } else {
+      // when the data is done being read in, then send a NACK to tell the
+      // slave to stop reading
+      write_bit(1, p_timeout);
+    }
+  }
+}
+
+hal::byte i2c_bit_bang::read_byte()
+{
+  constexpr auto byte_length = 8;
+  hal::byte read_byte = 0;
+  uint32_t i;
+  for (i = 1; i <= byte_length; i++) {
```

Comment:
Why is the iterator value outside of the for loop? We should try to keep variables within the lowest scope necessary.

Reply:
Missed this during refactoring, will change this.